### PR TITLE
Maintain consistency when deserializing to JSON

### PR DESCRIPTION
**Pull request by dhruvmanila**
## Summary

Maintain consistency while deserializing Jupyter notebook to JSON. The following
changes were made:

1. Use string array to store the source value as that's the default (https://github.com/jupyter/nbformat/blob/57817204230a8f7173b8bec380e571b5d410de0d/nbformat/v4/nbjson.py#L56-L57)
2. Remove unused structs and enums
3. Reorder the keys in alphabetical order as that's the default. (https://github.com/jupyter/nbformat/blob/57817204230a8f7173b8bec380e571b5d410de0d/nbformat/v4/nbjson.py#L51)

### Side effect

Removing the `preserve_order` feature means that the order of keys in JSON output (`--format json`) will be in alphabetical order. This is because the value is represented using `serde_json::Value` which internally is a `BTreeMap`, thus sorting it as per the string key. For posterity if this turns out to be not ideal, then we could define a struct representing the JSON object and the order of struct fields will determine the order in the JSON string.

## Test Plan

Add a test case to assert the raw JSON string.

---
***Edit: Both have been fixed, keeping here for posterity***

#### 1

`execution_count` is a required field _only_ in code cell while for others
it's not required. We use a single `Cell` struct and a `cell_type` field
instead of having structs for each cell type (`CodeCell`, `MarkdownCell`).
This means that the field `execution_count` will always be added to the JSON
string.

**What to do then?**

* We could either use a custom deserializer and check for the `cell_type`.
* Split the `Code` struct into 3 distinct types using enum:

```rust
enum CellType {
    RawCell(RawCell),
    MarkdownCell(MarkdownCell),
    CodeCell(CodeCell)
}

struct RawCell {
    ...
}

struct MarkdownCell {
    ...
}

struct CodeCell {
    ...
}
```

#### 2

For some fields, the `additionalProperties` is `true` which means there could be
unknown properties added which we've to keep while deserializing. We do this
using:

```rust
pub struct RawNotebookMetadata {
    // ... other fields

    #[serde(flatten)]
    pub other: BTreeMap<String, Value>
}
```

But, now the order won't be alphabetical. Wherever the `other` field is present,
all of the keys in it will be flattened at that position.

In the following example, the `nbconvert_exporter` and `version` keys are extra.
Now, as the `other` field in our struct is at the end, both the extra keys have
been moved to the end.

```diff
    "language_info": {
     "codemirror_mode": {
      "name": "ipython",
      "version": 3
     },
    "file_extension": ".py",
    "mimetype": "text/x-python",
    "name": "python",
-   "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
+   "nbconvert_exporter": "python",
    "version": "3.11.3"
   }
```

### CODE REVIEW

**Comment by dhruvmanila** [comment]
Current dependencies on/for this PR:
* main
  * **PR #5114** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5114" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/5114?utm_source=stack-comment).

**Comment by dhruvmanila** [comment]
Actually, we should resolve (1) as otherwise the validation fails:
```
NotebookValidationError: Additional properties are not allowed ('execution_count' was unexpected)

Failed validating 'additionalProperties' in markdown_cell:

On instance['cells'][0]:
{'cell_type': 'markdown',
 'execution_count': None,
 'id': 'f3c286e9-fa52-4440-816f-4449232f199a',
 'metadata': {},
 'source': '# Ruff Test'}
```

File: crates/ruff/src/jupyter/schema.rs [hunk 1]
```diff
@@ -49,110 +53,31 @@ pub struct RawNotebook {
 #[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
 #[serde(deny_unknown_fields)]
 pub struct Cell {
-    pub attachments: Option<HashMap<String, HashMap<String, Value>>>,
+    pub attachments: Option<Value>,
     /// String identifying the type of cell.
     pub cell_type: CellType,
+    /// The code cell's prompt number. Will be null if the cell has not been run.
+    #[serialize_always]
```
Comment (dhruvmanila) [inline crates/ruff/src/jupyter/schema.rs:60]: If we remove this, then we would get the following validation error:
```
NotebookValidationError: 'execution_count' is a required property

Failed validating 'required' in code_cell:

On instance['cells'][3]:
{'cell_type': 'code',
 'id': '6bced3f8-e0a4-450c-ae7c-f60ad5671ee9',
 'metadata': {},
 'outputs': ['...0 outputs...'],
 'source': 'with contextlib.suppress(ValueError):\n    print()\n'}
```

**Comment by konstin** [comment]
Re 1: My intution goes towards using an enum and `#[serde(flatten)]`, but both should do. Maybe you can put only the cell type and `execution_count` on each variant and tell serde to decide by cell type from a container attribute.

Re 2: That's a tricky problem! Maybe https://stackoverflow.com/a/67792465/3549270 would work?

**Comment by dhruvmanila** [comment]
> Re 2: That's a tricky problem! Maybe [stackoverflow.com/a/67792465/3549270](https://stackoverflow.com/a/67792465/3549270) would work?

I tried this, it didn't work. It just keeps the same order as defined in the struct which is the default behavior.

>  we add an extra "" at the end of list where jupyter doesn't,

Nice! Let me take a look at this.

**Comment by dhruvmanila** [comment]
@konstin Do you mean something like this? (I'm not sure how to use `#[serde(flatten)]` here)

```rust
#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[serde(tag = "cell_type")]
pub enum Cell {
    #[serde(rename = "code")]
    Code(CodeCell),
    #[serde(rename = "markdown")]
    Markdown(MarkdownCell),
    #[serde(rename = "raw")]
    Raw(RawCell),
}

#[skip_serializing_none]
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct RawCell {
    pub attachments: Option<Value>,
    pub id: Option<String>,
    pub metadata: Value,
    pub source: SourceValue,
}

/// Notebook markdown cell.
#[skip_serializing_none]
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct MarkdownCell {
    pub attachments: Option<Value>,
    pub id: Option<String>,
    pub metadata: Value,
    pub source: SourceValue,
}

/// Notebook code cell.
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct CodeCell {
    pub execution_count: Option<i64>,
    pub id: Option<String>,
    pub metadata: Value,
    pub outputs: Vec<Value>,
    pub source: SourceValue,
}
```

**Comment by konstin** [comment]
Like this (untested, but i've used flatten in other projects)
```rust
use serde::{Serialize, Deserialize};

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[serde(tag = "cell_type")]
pub enum Cell {
    #[serde(rename = "code")]
    Code {
        #[serde(flatten)]
        pub cell: CellInner,
        pub execution_count: Option<i64>,
    }
    #[serde(rename = "markdown")]
    Markdown {
        #[serde(flatten)]
        pub cell: CellInner,
    },
    #[serde(rename = "raw")]
    Raw {
        #[serde(flatten)]
        pub cell: CellInner,
    },
}

#[skip_serializing_none]
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct CellInner {
    pub attachments: Option<Value>,
    pub id: Option<String>,
    pub metadata: Value,
    pub source: SourceValue,
}
```

**Comment by konstin** [comment]
Could you add a notebook saved from the web interface of `jupyter notebook` and the same notebook after running it through `ruff --fix`, and then a test that checks that the string output of `ruff --fix` matches the saved fixed notebook? This way we can both see how the differences look in the repo and have a way to ensure that we don't accidentally change the json output. I'd compare the raw text and not the parsed json because that is what people will see in their git diffs.

**Comment by dhruvmanila** [comment]
I would argue to go with my implementation because:

* Using flatten will give us the same problem as mentioned in (2). The order of the fields will probably not be alphabetical.
	
	```json
    {
      "cell_type": "code",
      "id": "5e3ef98e-224c-450a-80e6-be442ad50907",
      "metadata": {
        "tags": []
      },
      "source": [
        "import contextlib\n",
        "import random"
      ],
      "outputs": [],
      "execution_count": 1
    }
	```
* There are other keys which has the same problem as `execution_count`
  * `outputs` is required only in code cells and not others
  * `attachments` is required only in raw and markdown cells and not the code cell

**Comment by dhruvmanila** [comment]
Oh, I found the reason why the sorting wasn't working earlier. We've declared `serde_json` with `preserve_order` feature which will preserve the insertion order and that's why the sorting was not done alphabetically.

**Comment by konstin** [comment]
> Using flatten will give us the same problem as mentioned in (2). The order of the fields will probably not be alphabetical.

That's a pity, yours is clearly the better solution then

> It seems like the feature was added during the initial PR to support Jupyter notebook (https://github.com/astral-sh/ruff/pull/3440/files#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542). Any reason why this was added and can we remove it?

none beyond "feels useful", feel free to remove it

**Comment by dhruvmanila** [comment]
## Update

* Update `Cell` schema to use tags to have independent structure for each cell type. This is to accommodate fields specific to the cell type.
* Sort the notebook fields in alphabetical order manually to allow any field order in the struct definition.
* Add a test case to assert JSON consistency (raw JSON, not parsed)
* Update output snapshots as the `preserve_order` feature got removed from `serde_json` dependency. The fields just got reordered.

### That trailing newline...

Well, the JSON string might contain a trailing newline which is handled by black: https://github.com/psf/black/blob/01b8d3d4095ebdb91d0d39012a517931625c63cb/src/black/__init__.py#L1024. Currently, we don't write the JSON string with a trailing newline. This needs to be handled.

**Comment by dhruvmanila** [comment]
@konstin I believe the integration test for Jupyter notebook here can be removed?

https://github.com/astral-sh/ruff/blob/4079f982f5634816808fd88c530ae848e5695f80/crates/ruff_cli/src/commands/run.rs#L190-L278

**Comment by konstin** [comment]
> I believe the integration test for Jupyter notebook here can be removed?

yep feel free to remove it

File: crates/ruff/src/jupyter/schema.rs [hunk 2]
```diff
@@ -5,33 +5,48 @@
...
+fn sort_alphabetically<T: Serialize, S: serde::Serializer>(
+    value: &T,
+    serializer: S,
+) -> Result<S::Ok, S::Error> {
+    let value = serde_json::to_value(value).map_err(serde::ser::Error::custom)?;
+    value.serialize(serializer)
+}
+
+#[derive(Serialize)]
+pub struct SortAlphabetically<T: Serialize>(#[serde(serialize_with = "sort_alphabetically")] pub T);
```
Comment (konstin) [inline crates/ruff/src/jupyter/schema.rs:36]: Could you add a comment why we need this indirection, i.e. why `sort_alphabetically` gives a sorted result where default serialization doesn't?

**Review by konstin — APPROVED** [review]
The test looks good!

**Pull request merged**